src/plugin/plugin.ts

```diff
@@ -32,6 +42,54 @@ export const createPlugin = <
 ): Plugin<PluginState, Schema> => {
   return new Plugin<PluginState, Schema>({
     key: pluginKey,
+    /**
+     * Update the elements to represent the current selection.
+     */
+    appendTransaction: (_, oldState, newState) => {
+      // If we are not transitioning between at least one selection of non-zero
+      // width, we cannot be altering the element selection state.
+      if (selectionIsZeroWidth(oldState) && selectionIsZeroWidth(newState)) {
+        return;
+      }
+
+      const tr = newState.tr;
+      const elementNodeToPos = new Map<Node, number>();
+
+      // Find all the nodes within the current selection.
+      newState.doc.nodesBetween(
+        newState.selection.from,
+        newState.selection.to,
+        (node, pos) => {
+          if (
+            isProseMirrorElement(node) &&
+            newState.selection.from <= pos &&
+            newState.selection.to >= pos + node.nodeSize
+          ) {
+            elementNodeToPos.set(node, pos);
+            return false;
+          }
+        }
+      );
+
+      // Update every relevant node with the new selection state.
+      newState.doc.descendants((node, pos) => {
+        const isCurrentlySelected = elementNodeToPos.get(node) !== undefined;
+        const shouldUpdateNode =
+          (isProseMirrorElementSelected(node) && !isCurrentlySelected) ||
+          (!isProseMirrorElementSelected(node) && isCurrentlySelected);
+
+        if (shouldUpdateNode) {
+          tr.setNodeMarkup(pos, undefined, {
+            ...node.attrs,
+            [elementSelectedNodeAttr]: isCurrentlySelected,
+          });
+
+          return false;
```

Comment:
A small possible hiccup, we are no longer descend nodes. This could potentially trip us up. We could wrap this descendants function in an `if(isProseMirrorElement)`, to resolve this.